File: back/DevC/src/main/java/com/back/devc/domain/post/post/service/PostService.java
package com.back.devc.domain.post.post.service;

import com.back.devc.domain.interaction.bookmark.repository.BookmarkRepository;
import com.back.devc.domain.interaction.postLike.repository.PostLikeRepository;
import com.back.devc.domain.member.member.entity.Member;
import com.back.devc.domain.member.member.repository.MemberRepository;
import com.back.devc.domain.member.searchLog.dto.CreateSearchLogRequest;
import com.back.devc.domain.member.searchLog.service.SearchLogService;
import com.back.devc.domain.post.category.entity.Category;
import com.back.devc.domain.post.category.repository.CategoryRepository;
import com.back.devc.domain.post.comment.repository.CommentRepository;
import com.back.devc.domain.post.post.dto.*;
import com.back.devc.domain.post.post.entity.Post;
import com.back.devc.domain.post.post.repository.PostRepository;
import com.back.devc.domain.post.post.type.PostSearchType;
import com.back.devc.domain.post.post.type.PostSortType;
import com.back.devc.global.exception.ApiException;
import com.back.devc.global.exception.errorCode.CategoryErrorCode;
import com.back.devc.global.exception.errorCode.MemberErrorCode;
import com.back.devc.global.exception.errorCode.PostErrorCode;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.lang.reflect.Field;

@Service
@RequiredArgsConstructor
@Transactional
public class PostService {

    private final PostRepository postRepository;
    private final MemberRepository memberRepository;
    private final CategoryRepository categoryRepository;
    private final PostLikeRepository postLikeRepository;
    private final BookmarkRepository bookmarkRepository;
    private final SearchLogService searchLogService;
    private final CommentRepository commentRepository;

    public PostCreateResponse write(Long userId, PostCreateRequest request) {

        Member member = memberRepository.findById(userId)
                .orElseThrow(() -> new ApiException(MemberErrorCode.MEMBER_NOT_FOUND));;

        Category category = categoryRepository.findById(request.categoryId())
                .orElseThrow(() -> new ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND));

        Post post = Post.builder()
                .member(member)
                .category(category)
                .title(request.title())
                .content(request.content())
                .build();

        Post saved = postRepository.save(post);

        return PostCreateResponse.from(saved);
    }

    @Transactional
    public PostDetailResponse findDetailById(Long postId, Long loginUserId) {
        Post post = postRepository.findByPostIdAndIsDeletedFalse(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        post.increaseViewCount();

        boolean liked = loginUserId != null
                && postLikeRepository.existsByMember_UserIdAndPost_PostId(loginUserId, postId);

        boolean bookmarked = loginUserId != null
                && bookmarkRepository.existsByMember_UserIdAndPost_PostId(loginUserId, postId);

        int bookmarkCount = Math.toIntExact(bookmarkRepository.countByPost_PostId(postId));
        applyActualCommentCount(post);

        return PostDetailResponse.from(post, liked, bookmarked, bookmarkCount);
    }

    @Transactional
    public Page<PostListResponse> getPosts(
            Long loginUserId,
            Long categoryId,
            String keyword,
            PostSearchType searchType,
            PostSortType sort,
            int page,
            int size
    ) {
        Sort jpaSort = switch (sort) {
            case VIEWS -> Sort.by(
                    Sort.Order.desc("viewCount"),
                    Sort.Order.desc("createdAt")
            );
            case LIKES -> Sort.by(
                    Sort.Order.desc("likeCount"),
                    Sort.Order.desc("createdAt")
            );
            default -> Sort.by(Sort.Order.desc("createdAt"));
        };

        Pageable pageable = PageRequest.of(page, size, jpaSort);

        boolean hasKeyword = keyword != null && !keyword.isBlank();

        if (loginUserId != null && hasKeyword) {
            searchLogService.createSearchLog(
                    loginUserId,
                    new CreateSearchLogRequest(keyword.trim())
            );
        }

        Page<Post> result;

        if (hasKeyword) {
            String kw = keyword.trim();

            if (searchType == PostSearchType.TITLE) {
                result = (categoryId != null)
                        ? postRepository.findByCategoryCategoryIdAndTitleContainingAndIsDeletedFalse(categoryId, kw, pageable)
                        : postRepository.findByTitleContainingAndIsDeletedFalse(kw, pageable);
            } else if (searchType == PostSearchType.CONTENT) {
                result = (categoryId != null)
                        ? postRepository.findByCategoryCategoryIdAndContentContainingAndIsDeletedFalse(categoryId, kw, pageable)
                        : postRepository.findByContentContainingAndIsDeletedFalse(kw, pageable);
            } else {
                result = (categoryId != null)
                        ? postRepository.searchPosts(categoryId, kw, pageable)
                        : postRepository.searchByKeyword(kw, pageable);
            }
        } else {
            if (categoryId != null) {
                if (!categoryRepository.existsById(categoryId)) {
                    throw new ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND);
                }
                result = postRepository.findByCategoryCategoryIdAndIsDeletedFalse(categoryId, pageable);
            } else {
                result = postRepository.findByIsDeletedFalse(pageable);
            }
        }

        return result.map(post -> {
            applyActualCommentCount(post);
            boolean liked = loginUserId != null
                    && postLikeRepository.existsByMember_UserIdAndPost_PostId(loginUserId, post.getPostId());

            boolean bookmarked = loginUserId != null
                    && bookmarkRepository.existsByMember_UserIdAndPost_PostId(loginUserId, post.getPostId());

            return PostListResponse.from(post, liked, bookmarked);
        });
    }

    private void applyActualCommentCount(Post post) {
        int actualCommentCount = Math.toIntExact(commentRepository.countByPostIdAndIsDeletedFalse(post.getPostId()));

        try {
            Field commentCountField = Post.class.getDeclaredField("commentCount");
            commentCountField.setAccessible(true);
            commentCountField.setInt(post, actualCommentCount);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new IllegalStateException("Post commentCount 동기화에 실패했습니다.", e);
        }
    }

    @Transactional
    public void increaseViewCount(Long postId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        post.increaseViewCount();
    }

    @Transactional
    public void increaseCommentCount(Long postId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        post.increaseCommentCount();
    }

    @Transactional
    public void decreaseCommentCount(Long postId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        post.decreaseCommentCount();
    }

    public PostUpdateResponse update(Long memberId, Long postId, PostUpdateRequest request) {

        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        if (!post.getMember().getUserId().equals(memberId)) {
            throw new ApiException(PostErrorCode.POST_403_FORBIDDEN);
        }

        if (post.isDeleted()) {
            throw new ApiException(PostErrorCode.POST_401_1_ALREADY_DELETED);
        }

        Category category = categoryRepository.findById(request.categoryId())
                .orElseThrow(() -> new ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND));

        post.update(
                request.title(),
                request.content(),
                category
        );

        return PostUpdateResponse.from(post);
    }

    public PostDeleteResponse delete(Long memberId, Long postId) {

        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ApiException(PostErrorCode.POST_404_NOT_FOUND));

        if (!post.getMember().getUserId().equals(memberId)) {
            throw new ApiException(PostErrorCode.POST_403_FORBIDDEN);
        }

        if (post.isDeleted()) {
            throw new ApiException(PostErrorCode.POST_401_1_ALREADY_DELETED);
        }

        bookmarkRepository.deleteByPost_PostId(postId);
        postLikeRepository.deleteByPost_PostId(postId);

        post.delete();

        return PostDeleteResponse.of(postId);
    }
}

File: back/DevC/src/main/kotlin/com/back/devc/domain/post/post/service/PostService.kt
package com.back.devc.domain.post.post.service

import com.back.devc.domain.interaction.bookmark.repository.BookmarkRepository
import com.back.devc.domain.interaction.postLike.repository.PostLikeRepository
import com.back.devc.domain.member.member.repository.MemberRepository
import com.back.devc.domain.member.searchLog.dto.CreateSearchLogRequest
import com.back.devc.domain.member.searchLog.service.SearchLogService
import com.back.devc.domain.post.category.repository.CategoryRepository
import com.back.devc.domain.post.comment.repository.CommentRepository
import com.back.devc.domain.post.post.dto.PostCreateRequest
import com.back.devc.domain.post.post.dto.PostCreateResponse
import com.back.devc.domain.post.post.dto.PostDeleteResponse
import com.back.devc.domain.post.post.dto.PostDetailResponse
import com.back.devc.domain.post.post.dto.PostListResponse
import com.back.devc.domain.post.post.dto.PostUpdateRequest
import com.back.devc.domain.post.post.dto.PostUpdateResponse
import com.back.devc.domain.post.post.entity.Post
import com.back.devc.domain.post.post.repository.PostRepository
import com.back.devc.domain.post.post.type.PostSearchType
import com.back.devc.domain.post.post.type.PostSortType
import com.back.devc.global.exception.ApiException
import com.back.devc.global.exception.errorCode.CategoryErrorCode
import com.back.devc.global.exception.errorCode.MemberErrorCode
import com.back.devc.global.exception.errorCode.PostErrorCode
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class PostService(
    private val postRepository: PostRepository,
    private val memberRepository: MemberRepository,
    private val categoryRepository: CategoryRepository,
    private val postLikeRepository: PostLikeRepository,
    private val bookmarkRepository: BookmarkRepository,
    private val searchLogService: SearchLogService,
    private val commentRepository: CommentRepository,
) {

    @Transactional
    fun write(
        userId: Long,
        request: PostCreateRequest,
    ): PostCreateResponse {
        val member = memberRepository.findById(userId)
            .orElseThrow { ApiException(MemberErrorCode.MEMBER_NOT_FOUND) }

        val category = categoryRepository.findById(request.categoryId)
            .orElseThrow { ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND) }

        val post = Post.Companion.create(
            member = member,
            category = category,
            title = request.title,
            content = request.content,
        )

        val savedPost = postRepository.save(post)


        return PostCreateResponse.Companion.from(savedPost)
    }

    @Transactional
    fun findDetailById(
        postId: Long,
        loginUserId: Long?,
    ): PostDetailResponse {
        val post = findActivePost(postId)

        post.increaseViewCount()

        val liked = loginUserId?.let {
            postLikeRepository.existsByMember_UserIdAndPost_PostId(it, postId)
        } ?: false

        val bookmarked = loginUserId?.let {
            bookmarkRepository.existsByMember_UserIdAndPost_PostId(it, postId)
        } ?: false

        val bookmarkCount = bookmarkRepository.countByPost_PostId(postId).toInt()

        syncCommentCount(post)

        return PostDetailResponse.Companion.from(
            post,
            liked,
            bookmarked,
            bookmarkCount,
        )
    }

    fun getPosts(
        loginUserId: Long?,
        categoryId: Long?,
        keyword: String?,
        searchType: PostSearchType?,
        sort: PostSortType,
        page: Int,
        size: Int,
    ): Page<PostListResponse> {
        val pageable: Pageable = PageRequest.of(
            page,
            size,
            createSort(sort),
        )

        val trimmedKeyword = keyword?.trim()?.takeIf { it.isNotBlank() }

        if (loginUserId != null && trimmedKeyword != null) {
            searchLogService.createSearchLog(
                loginUserId,
                CreateSearchLogRequest(trimmedKeyword),
            )
        }

        val posts = searchPosts(
            categoryId = categoryId,
            keyword = trimmedKeyword,
            searchType = searchType,
            pageable = pageable,
        )

        return posts.map { post ->

            val postId = requireNotNull(post.postId)

            val liked = loginUserId?.let {
                postLikeRepository.existsByMember_UserIdAndPost_PostId(it, postId)
            } ?: false

            val bookmarked = loginUserId?.let {
                bookmarkRepository.existsByMember_UserIdAndPost_PostId(it, postId)
            } ?: false

            PostListResponse.Companion.from(
                post,
                liked,
                bookmarked,
            )
        }
    }

    @Transactional
    fun update(
        memberId: Long,
        postId: Long,
        request: PostUpdateRequest,
    ): PostUpdateResponse {
        val post = findActivePost(postId)

        validatePostOwner(post, memberId)

        val category = categoryRepository.findById(request.categoryId)
            .orElseThrow { ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND) }

        post.update(
            request.title,
            request.content,
            category,
        )

        return PostUpdateResponse.Companion.from(post)
    }

    @Transactional
    fun delete(
        memberId: Long,
        postId: Long,
    ): PostDeleteResponse {
        val post = findActivePost(postId)

        validatePostOwner(post, memberId)

        bookmarkRepository.deleteByPost_PostId(postId)
        postLikeRepository.deleteByPost_PostId(postId)

        commentRepository.findByPostIdAndIsDeletedFalse(postId)
            .forEach { comment ->
                comment.softDelete()
            }

        post.delete()

        return PostDeleteResponse.Companion.of(postId)
    }

    @Transactional
    fun increaseViewCount(postId: Long) {
        val post = findActivePost(postId)
        post.increaseViewCount()
    }

    @Transactional
    fun increaseCommentCount(postId: Long) {
        val post = findActivePost(postId)
        post.increaseCommentCount()
    }

    @Transactional
    fun decreaseCommentCount(postId: Long) {
        val post = findActivePost(postId)
        post.decreaseCommentCount()
    }

    private fun searchPosts(
        categoryId: Long?,
        keyword: String?,
        searchType: PostSearchType?,
        pageable: Pageable,
    ): Page<Post> {
        if (keyword != null) {
            return when (searchType) {
                PostSearchType.TITLE -> {
                    if (categoryId != null) {
                        postRepository.findByCategoryCategoryIdAndTitleContainingAndIsDeletedFalse(
                            categoryId,
                            keyword,
                            pageable,
                        )
                    } else {
                        postRepository.findByTitleContainingAndIsDeletedFalse(
                            keyword,
                            pageable,
                        )
                    }
                }

                PostSearchType.CONTENT -> {
                    if (categoryId != null) {
                        postRepository.findByCategoryCategoryIdAndContentContainingAndIsDeletedFalse(
                            categoryId,
                            keyword,
                            pageable,
                        )
                    } else {
                        postRepository.findByContentContainingAndIsDeletedFalse(
                            keyword,
                            pageable,
                        )
                    }
                }

                else -> {
                    if (categoryId != null) {
                        postRepository.searchPosts(
                            categoryId,
                            keyword,
                            pageable,
                        )
                    } else {
                        postRepository.searchByKeyword(
                            keyword,
                            pageable,
                        )
                    }
                }
            }
        }

        if (categoryId != null) {
            if (!categoryRepository.existsById(categoryId)) {
                throw ApiException(CategoryErrorCode.CATEGORY_404_NOT_FOUND)
            }

            return postRepository.findByCategoryCategoryIdAndIsDeletedFalse(
                categoryId,
                pageable,
            )
        }

        return postRepository.findByIsDeletedFalse(pageable)
    }

    private fun createSort(sort: PostSortType): Sort {
        return when (sort) {
            PostSortType.VIEWS -> Sort.by(
                Sort.Order.desc("viewCount"),
                Sort.Order.desc("createdAt"),
            )

            PostSortType.LIKES -> Sort.by(
                Sort.Order.desc("likeCount"),
                Sort.Order.desc("createdAt"),
            )

            else -> Sort.by(
                Sort.Order.desc("createdAt"),
            )
        }
    }

    private fun findActivePost(postId: Long): Post {
        return postRepository.findByPostIdAndIsDeletedFalse(postId)
            .orElseThrow {
                ApiException(PostErrorCode.POST_404_NOT_FOUND)
            }
    }

    private fun validatePostOwner(
        post: Post,
        memberId: Long,
    ) {
        if (post.member.userId != memberId) {
            throw ApiException(PostErrorCode.POST_403_FORBIDDEN)
        }

        if (post.isDeleted) {
            throw ApiException(PostErrorCode.POST_401_1_ALREADY_DELETED)
        }
    }

    private fun syncCommentCount(post: Post) {
        val postId = requireNotNull(post.postId)

        val actualCommentCount = commentRepository
            .countByPostIdAndIsDeletedFalse(postId)
            .toInt()
    }
}